cluster-default-publishing-strategy.md

```diff
@@ -0,0 +1,138 @@
+
+This should not block administrators from using operator-level configuration to override publishing strategy of individual endpoints.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+See the [section][#cluster-publishing-strategy]
+
+## Motivation
+
+See the [section][#cluster-publishing-strategy]
+
+### Goals
+
+1. Define an API that allows all the cluster operators to discover the default cluster publishing strategy.
+
+### Non-Goals
+
+## Proposal
+
+Modify the existing `config.openshift.io`'s custom resource [`Infrastructure`][infrastructure-api] to include an enum field `Publish` that can be set to `External` or `Internal`.
+
+The field will be added to [`status` section][infrastructure-api-status] because the goal is to provide an API for discovery i.e. the current state, and not to provide an API for customers to specify the desired state.
+
+### User Stories
+
+#### Default Ingress Controller LoadBalancer Scope
+
+The ingress-operator will be able to use this field to define the **default** [scope][ingress-controller-api-load-balancer-scope] for the LoadBalancers for the [ingress controllers][ingress-controller-api].
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Addition of the new field requires migration of all the existing clusters. Since we have only supported `External` strategy for OpenShift, the migration should be simple enough.
+
+## Design Details
+
+### New API
+
+`PublishingStrategy` enum definition.
+
+```go
+// PublishingStrategy is a strategy for how various endpoints for the cluster are exposed.
+type PublishingStrategy string
+
+const (
+    // ExternalPublishingStrategy exposes endpoints for the cluster to the Internet
+    ExternalPublishingStrategy PublishingStrategy = “External”
+    // InternalPublishingStrategy exposes the endpoints for the cluster to the private network only.
+    InternalPublishingStrategy PublishingStrategy = “Internal”
+)
+```
+
+Updating the `InfrastructureStatus` object to add `Publish` field.
+
+```go
+type InfrastructureStatus struct {
+...
+
+// Publish controls how the user facing endpoints of the cluster like the k8s API, OpenShift routes etc. are exposed.
+// When no strategy is specified, the strategy is `External`.
+// +optional
+Publish PublishingStrategy `json:”publish,omitempty”`
+
+...
+
+}
+```
+
+### Migration of existing clusters
+
+All clusters with `empty` PublishingStrategy can be updated to `External`, while all non-empty values will be kept untouched.
+
+#### Rollbacks
+
+In case of rollbacks after the migration has been done, the value `External` will be accurate.
+
+### Test Plan
+
```

Comment:
how would this be tested?

Reply:
@deads2k do you mean?

a) test if this API is set correctly?
b) Or if the migration is happening correctly?

Reply:
> @deads2k do you mean?
> 
> a) test if this API is set correctly?
> b) Or if the migration is happening correctly?

This feature doesn't exist in a vacuum.  It's not really useful unless something is done with this field.  I'm asking how you would ensure that your initial consumers worked correctly.